src/Database/Redis/Connection.hs

```diff
@@ -294,17 +294,17 @@ refreshShardMap connectInfo@ConnInfo{connectMaxConnections,connectMaxIdleTime} (
                         return $ HM.insert nodeid nodeConnPool acc
                  ) HM.empty (nub $ Cluster.nodes newShardMap)
 
-refreshShardMapWithNodeConn :: [Cluster.NodeConnection] -> IO ShardMap
-refreshShardMapWithNodeConn [] = throwIO $ ClusterConnectError (Error "Couldn't refresh shardMap due to connection error")
-refreshShardMapWithNodeConn nodeConnsList = do
+refreshShardMapWithNodeConn :: Maybe Cluster.NodeConnection -> [Cluster.NodeConnection] -> IO ShardMap
+refreshShardMapWithNodeConn Nothing [] = throwIO $ ClusterConnectError (Error "Couldn't refresh shardMap due to connection error")
+refreshShardMapWithNodeConn maybeNodeConn nodeConnsList = do
     let numOfNodes = length nodeConnsList
     selectedIdx <- randomRIO (0, length nodeConnsList - 1)
-    let (Cluster.NodeConnection pool _) = nodeConnsList !! selectedIdx
+    let (Cluster.NodeConnection pool _) = fromMaybe (nodeConnsList !! selectedIdx) maybeNodeConn
     eresp <- try $ refreshShardMapWithPool pool
     case eresp of 
         Left  (_::SomeException) ->  do                 -- retry on other node
             let otherSelectedIdx                        = (selectedIdx + 1) `mod` numOfNodes
-                (Cluster.NodeConnection otherPool _)    = nodeConnsList !! otherSelectedIdx
+                (Cluster.NodeConnection otherPool _)    = fromMaybe (nodeConnsList !! otherSelectedIdx) (maybeNodeConn >>= \nodeConn -> find (nodeConn /=) nodeConnsList)
             refreshShardMapWithPool otherPool
         Right shardMap -> return shardMap
     where 
```

Comment:
> Left (err :: SomeException) -> do
>                                     print $ "ShardMapRefreshError-" <> show err 
>                                     throwIO $ ClusterConnectE

rror (Error "Couldn't refresh shardMap due to connection error")
                                    
Can you throw this `err` as part of throwIO ?